File: evals/runner/score.py
"""Prototype eval scorer: the deterministic parts of evals/scoring.md.

  python3 evals/runner/score.py seeded <case_dir> <run_dir>          # §A finding matching
  python3 evals/runner/score.py golden <case_dir> <run_dir>          # §B alignment proxy
  python3 evals/runner/score.py judge  <case_dir> <judge_output.json> # judge verdicts vs expected

Each command prints a Markdown report and writes <run_dir or output dir>/score.json.
Release thresholds come from evals/thresholds.yaml. Semantic decisions (judge) are marked 'unconfirmed'.
Needs: pyyaml, jsonschema.
"""
import json, pathlib, re, sys, yaml, jsonschema

EVALS = pathlib.Path(__file__).resolve().parents[1]
THRESH = yaml.safe_load((EVALS / "thresholds.yaml").read_text())
SEV = {"minor": 1, "major": 2, "blocker": 3}
PRIORITY_W = {"must": 3, "should": 2, "could": 1, "wont": 0}

# ---------------------------------------------------------------- parsing (same grammar as spec-lint)
def parse_md(path):
    txt = pathlib.Path(path).read_text()
    m = re.match(r"^---\n(.*?)\n---\n(.*)$", txt, re.S)
    hdr, body = yaml.safe_load(m.group(1)), re.sub(r"<!--.*?-->", "", m.group(2), flags=re.S)
    blocks, cur, last = {}, None, None
    for line in body.splitlines():
        if line.startswith("## "):
            cur = None; continue
        mb = re.match(r"^### ([A-Z]+-[0-9.]+) · (.+)$", line)
        if mb:
            cur = blocks[mb.group(1)] = {"_id": mb.group(1), "_title": mb.group(2)}; continue
        if cur is None or not line.strip(): continue
        mk = re.match(r"^- ([a-z_]+):\s?(.*)$", line)
        if mk:
            v = mk.group(2).strip()
            cur[mk.group(1)] = [x.strip() for x in v[1:-1].split(",") if x.strip()] if v.startswith("[") else (v or [])
            last = mk.group(1); continue
        mi = re.match(r"^  - (.+)$", line)
        if mi and isinstance(cur.get(last), list): cur[last].append(mi.group(1))
    return hdr, blocks

def links(run_dir):
    return (yaml.safe_load((pathlib.Path(run_dir) / "trace.yaml").read_text()) or {}).get("links") or []

def verdict(value, limit, kind):
    ok = value >= limit if kind == "min" else value <= limit
    return "✓" if ok else "✕"

# ---------------------------------------------------------------- §A seeded
def score_seeded(case_dir, run_dir):
    case = yaml.safe_load((case_dir / "case.yaml").read_text())
    exp = yaml.safe_load((case_dir / case["expected"]["findings"]).read_text())["findings"]
    step = case["targets"][0]
    art = "requirements-analysis.md" if step == "req-analysis" else "design-analysis.md"
    _, blocks = parse_md(run_dir / art)
    L = links(run_dir)
    actual = {bid: dict(b, affects={l["to"] for l in L if l["from"] == bid and l["rel"] == "affects"})
              for bid, b in blocks.items() if bid[:3] in ("RAF", "DAF") and "withdrawn" not in b}

    def quality(ef, af):
        return (0.5 if af["type"] in ef["types"] else 0) + \
               (0.3 if SEV[af["severity"]] >= SEV[ef["min_severity"]] else 0) + \
               (0.2 if af["route"] in ef["routes"] else 0)
    pairs = sorted(((quality(ef, af), -int(re.sub(r"\D", "", aid)), ef["id"], aid)
                    for ef in exp for aid, af in actual.items() if af["affects"] & set(ef["affects_any"])), reverse=True)
    used_e, used_a, match = set(), set(), {}
    for q, _, eid, aid in pairs:
        if eid in used_e or aid in used_a: continue
        used_e.add(eid); used_a.add(aid); match[eid] = aid

    rows, credit, blk_n, blk_hit, sev_ok, route_ok = [], 0.0, 0, 0, 0, 0
    for ef in exp:
        aid = match.get(ef["id"])
        af = actual.get(aid)
        if ef["min_severity"] == "blocker": blk_n += 1
        if not af:
            rows.append((ef["id"], "—", "missed", "", "", ef["about"])); continue
        t_ok = af["type"] in ef["types"]
        s_ok = SEV[af["severity"]] >= SEV[ef["min_severity"]]
        r_ok = af["route"] in ef["routes"]
        credit += 1.0 if t_ok else 0.5
        sev_ok += s_ok; route_ok += r_ok
        if ef["min_severity"] == "blocker" and s_ok: blk_hit += 1
        status = "matched" if t_ok else "found-mislabelled (unconfirmed)"
        rows.append((ef["id"], aid, status,
                     f'{af["severity"]}{"" if s_ok else " ✕ (≥ " + ef["min_severity"] + ")"}',
                     f'{af["route"]}{"" if r_ok else " ✕ " + "/".join(ef["routes"])}', ef["about"]))
    n_m = len(match)
    extras = [aid for aid in actual if aid not in used_a]
    m = {
        "recall.blocker": round(blk_hit / blk_n, 3) if blk_n else 1.0,
        "recall.all": round(credit / len(exp), 3),
        "severity_accuracy": round(sev_ok / n_m, 3) if n_m else 0.0,
        "route_accuracy": round(route_ok / n_m, 3) if n_m else 0.0,
        "noise": round(len(extras) / len(actual), 3) if actual else 0.0,
    }
    th = THRESH["release"][step]["seeded"]
    checks = [(k, m[k], th[k], "min") for k in ("recall.blocker", "recall.all", "severity_accuracy", "route_accuracy")] + \
             [("noise", m["noise"], th["noise_max"], "max")]
    out = [f"# Seeded score · {case['id']} · run `{run_dir.name}`", "",
           "| Expected | Actual | Result | Severity | Route | About |", "|---|---|---|---|---|---|"]
    out += [f"| {a} | {b} | {c} | {d} | {e} | {f} |" for a, b, c, d, e, f in rows]
    out += ["", f"Unmatched actual findings (label each `valid-extra` or `noise`): {', '.join(extras) or 'none'}", "",
            f"| Metric | Value | Release ({step}) |", "|---|---|---|"]
    out += [f"| `{k}` | {v} | {'≥' if kind == 'min' else '≤'} {lim} {verdict(v, lim, kind)} |" for k, v, lim, kind in checks]
    passed = all(verdict(v, lim, kind) == "✓" for _, v, lim, kind in checks)
    out += ["", f"**Case result: {'PASS' if passed else 'FAIL'}** (single run; release uses the worst of k={THRESH['release']['k_runs']} for hard rules)"]
    return {"case": case["id"], "run": run_dir.name, "metrics": m, "matches": match, "extras": extras, "pass": passed}, "\n".join(out)

# ---------------------------------------------------------------- §B golden (proxy)
def score_golden(case_dir, run_dir):
    case = yaml.safe_load((case_dir / "case.yaml").read_text())
    ref_dir = (case_dir / case["expected"]["reference"]).resolve()
    _, R = parse_md(ref_dir / "requirements.md")
    _, A = parse_md(run_dir / "requirements.md")
    R = {k: v for k, v in R.items() if k.startswith("REQ") and "withdrawn" not in v}
    A = {k: v for k, v in A.items() if k.startswith("REQ") and "withdrawn" not in v}
    jac = lambda a, b: len(set(a) & set(b)) / len(set(a) | set(b)) if (a or b) else 0.0
    cand = sorted(((jac(r["terms"], a["terms"]), rid, aid) for rid, r in R.items() for aid, a in A.items()
                   if r["context"] == a["context"] and jac(r["terms"], a["terms"]) >= 0.4),
                  key=lambda x: (-(x[0] >= 0.6), -x[0], x[1], x[2]))
    used_r, used_a, res = set(), set(), {}
    for j, rid, aid in cand:                                   # one-to-one, equivalents first
        if rid in used_r or aid in used_a: continue
        used_r.add(rid); used_a.add(aid)
        res[rid] = (aid, "equivalent" if j >= 0.6 else "partial", 1.0 if j >= 0.6 else 0.5, j)
    for j, rid, aid in cand:                                   # merges
        if rid in used_r or j < 0.6: continue
        used_r.add(rid); res[rid] = (aid, "merged", 0.5, j)
    num = sum(PRIORITY_W[R[r]["priority"]] * res.get(r, (None, None, 0.0))[2] for r in R)
    den = sum(PRIORITY_W[R[r]["priority"]] for r in R)
    must_miss = [r for r in R if R[r]["priority"] == "must" and r not in res]
    extras = [a for a in A if a not in used_a]
    m = {"ref_coverage": round(num / den, 3), "must_miss": len(must_miss)}
    th = THRESH["release"]["req-generation"]["golden"]
    checks = [("ref_coverage", m["ref_coverage"], th["ref_coverage"], "min"), ("must_miss", m["must_miss"], th["must_miss"], "max")]
    out = [f"# Golden score (proxy) · {case['id']} · run `{run_dir.name}`", "",
           "| Reference | Priority | Run | Decision (unconfirmed) | Term Jaccard |", "|---|---|---|---|---|"]
    for rid, r in R.items():
        aid, dec, _, j = res.get(rid, ("—", "missed", 0, 0))
        out.append(f"| {rid} · {r['_title']} | {r['priority']} | {aid} | {dec} | {j:.2f} |")
    extra_txt = ", ".join(a + " · " + A[a]["_title"] for a in extras) or "none"
    out += ["", f"Run REQs with no reference match (review: `valid-extra` / `scope-creep` / `duplicate`): {extra_txt}", "",
            "| Metric | Value | Release (req-generation) |", "|---|---|---|"]
    out += [f"| `{k}` | {v} | {'≥' if kind == 'min' else '≤'} {lim} {verdict(v, lim, kind)} |" for k, v, lim, kind in checks]
    passed = all(verdict(v, lim, kind) == "✓" for _, v, lim, kind in checks)
    out += ["", f"**Case result (proxy): {'PASS' if passed else 'FAIL'}**. The judge confirms or overturns each decision, then scores `ac_coverage`."]
    return {"case": case["id"], "run": run_dir.name, "metrics": m, "alignment": {k: v[:2] for k, v in res.items()},
            "must_missed": must_miss, "extras": extras, "pass": passed, "proxy": True}, "\n".join(out)

# ---------------------------------------------------------------- judge fixture check
def score_judge(case_dir, judge_json):
    case = yaml.safe_load((case_dir / "case.yaml").read_text())
    out_json = json.loads(pathlib.Path(judge_json).read_text())
    jsonschema.validate(out_json, json.loads((EVALS / "schemas" / "judge-output.schema.json").read_text()))
    got = {c["id"]: c["score"] for c in out_json["criteria"]}
    rows, ok_n = [], 0
    for ev in case["expected"]["judge_verdicts"]:
        s = got.get(ev["criterion"])
        ok = s is not None and ev.get("min_score", 1) <= s <= ev["max_score"]
        ok_n += ok
        rows.append(f"| {ev['criterion']} | {ev.get('min_score',1)}–{ev['max_score']} | {s if s is not None else 'missing'} | {'✓' if ok else '✕'} | {ev.get('why','')} |")
    v_ok = out_json["verdict"] == case["expected"]["verdict"]
    acc = round((ok_n + v_ok) / (len(rows) + 1), 3)
    out = [f"# Judge fixture · {case['id']}", "", "| Criterion | Expected | Judge | OK | Why |", "|---|---|---|---|---|", *rows, "",
           f"Verdict: expected **{case['expected']['verdict']}**, judge said **{out_json['verdict']}** {'✓' if v_ok else '✕'}", "",
           f"`verdict_accuracy` = {acc} (release ≥ {THRESH['release']['eval-judge']['judge']['verdict_accuracy']})"]
    return {"case": case["id"], "verdict_accuracy": acc, "verdict_ok": v_ok}, "\n".join(out)

if __name__ == "__main__":
    mode, case_dir, target = sys.argv[1], pathlib.Path(sys.argv[2]), pathlib.Path(sys.argv[3])
    fn = {"seeded": score_seeded, "golden": score_golden, "judge": score_judge}[mode]
    data, report = fn(case_dir, target)
    dest = (target if target.is_dir() else target.parent) / "score.json"
    dest.write_text(json.dumps(data, indent=2, ensure_ascii=False))
    print(report)

File: packages/eval-runner/src/score.ts
/**
 * The deterministic parts of evals/scoring.md, ported from evals/runner/score.py:
 *   seeded: §A finding matching · golden: §B alignment proxy · judge: judge scores vs a case's expected verdicts.
 * Each returns score data and the same Markdown report as the prototype, byte for byte.
 * Semantic decisions (the judge's) are marked 'unconfirmed'.
 */
import { readFileSync } from "node:fs";
import { basename, join, resolve } from "node:path";
import { Ajv2020 } from "ajv/dist/2020.js";
import { parse as parseYaml } from "yaml";
import { Diagnostics, fieldList, fieldScalar, parseMarkdown, type Block } from "spec-lint";

export const EVALS = resolve(import.meta.dirname, "..", "..", "..", "evals");
const SEV: Record<string, number> = { minor: 1, major: 2, blocker: 3 };
const PRIORITY_W: Record<string, number> = { must: 3, should: 2, could: 1, wont: 0 };
/** Metrics that are counts. Every other metric and threshold is a ratio. */
const COUNTS = new Set(["must_miss"]);

type Yaml = Record<string, any>;
export type Kind = "min" | "max";
export interface Check {
  metric: string;
  value: number;
  limit: number;
  kind: Kind;
}

export function loadThresholds(evals = EVALS): Yaml {
  return parseYaml(readFileSync(join(evals, "thresholds.yaml"), "utf8"));
}

function loadCaseYaml(caseDir: string): Yaml {
  return parseYaml(readFileSync(join(caseDir, "case.yaml"), "utf8"));
}

/** A number as Python's str() prints it: ratios keep a ".0" when whole, counts don't. */
export function num(v: number, metric = ""): string {
  return !COUNTS.has(metric) && Number.isInteger(v) ? v.toFixed(1) : String(v);
}

const round3 = (x: number) => Math.round(x * 1000) / 1000;
const ok = (c: Check) => (c.kind === "min" ? c.value >= c.limit : c.value <= c.limit);
const mark = (c: Check) => (ok(c) ? "✓" : "✕");

function checkRows(checks: Check[]): string[] {
  return checks.map((c) => `| \`${c.metric}\` | ${num(c.value, c.metric)} | ${c.kind === "min" ? "≥" : "≤"} ${num(c.limit, c.metric)} ${mark(c)} |`);
}

interface Artifact {
  blocks: Map<string, Block>;
}

function parseArtifact(path: string): Artifact {
  const md = parseMarkdown(readFileSync(path, "utf8"), basename(path), new Diagnostics());
  return { blocks: new Map(md.blocks.map((b) => [b.id, b])) };
}

const live = (b: Block) => !b.fields.has("withdrawn");

function traceLinks(specDir: string): Array<{ from: string; rel: string; to: string }> {
  return (parseYaml(readFileSync(join(specDir, "trace.yaml"), "utf8")) as Yaml)?.links ?? [];
}

// ---------------------------------------------------------------- §A seeded

export interface SeededScore {
  case: string;
  run: string;
  metrics: Record<"recall.blocker" | "recall.all" | "severity_accuracy" | "route_accuracy" | "noise", number>;
  /** Expected finding ID → the actual finding matched to it. */
  matches: Record<string, string>;
  /** Actual findings that match no expected finding. */
  extras: string[];
  pass: boolean;
}

interface ExpectedFinding {
  id: string;
  about: string;
  types: string[];
  affects_any: string[];
  min_severity: string;
  routes: string[];
}

/** Checks for the `seeded` suite of a step against thresholds.yaml → release. */
export function seededChecks(m: SeededScore["metrics"], th: Yaml): Check[] {
  return [
    ...(["recall.blocker", "recall.all", "severity_accuracy", "route_accuracy"] as const).map((k) => ({ metric: k, value: m[k], limit: th[k], kind: "min" as Kind })),
    { metric: "noise", value: m.noise, limit: th.noise_max, kind: "max" },
  ];
}

export function scoreSeeded(caseDir: string, specDir: string, runName = basename(specDir), thresholds = loadThresholds()): { data: SeededScore; report: string } {
  const c = loadCaseYaml(caseDir);
  const exp = (parseYaml(readFileSync(join(caseDir, c.expected.findings), "utf8")) as Yaml).findings as ExpectedFinding[];
  const step: string = c.targets[0];
  const art = step === "req-analysis" ? "requirements-analysis.md" : "design-analysis.md";
  const { blocks } = parseArtifact(join(specDir, art));
  const links = traceLinks(specDir);
  const actual = new Map(
    [...blocks.values()]
      .filter((b) => (b.prefix === "RAF" || b.prefix === "DAF") && live(b))
      .map((b) => [
        b.id,
        {
          type: fieldScalar(b, "type") ?? "",
          severity: fieldScalar(b, "severity") ?? "",
          route: fieldScalar(b, "route") ?? "",
          affects: new Set(links.filter((l) => l.from === b.id && l.rel === "affects").map((l) => l.to)),
        },
      ]),
  );
  type AF = NonNullable<ReturnType<typeof actual.get>>;
  const sev = (s: string) => SEV[s] ?? 0;
  const tOk = (ef: ExpectedFinding, af: AF) => ef.types.includes(af.type);
  const sOk = (ef: ExpectedFinding, af: AF) => sev(af.severity) >= sev(ef.min_severity);
  const rOk = (ef: ExpectedFinding, af: AF) => ef.routes.includes(af.route);
  const quality = (ef: ExpectedFinding, af: AF) => (tOk(ef, af) ? 0.5 : 0) + (sOk(ef, af) ? 0.3 : 0) + (rOk(ef, af) ? 0.2 : 0);

  // Highest quality first; ties go to the lower AF number (then, as in the prototype, the higher EF and AF IDs).
  const cmp = (a: string, b: string) => (a < b ? -1 : a > b ? 1 : 0);
  const afNum = (id: string) => Number(id.replace(/\D/g, ""));
  const pairs = exp
    .flatMap((ef) => [...actual].filter(([, af]) => ef.affects_any.some((t) => af.affects.has(t))).map(([aid, af]) => ({ q: quality(ef, af), n: afNum(aid), eid: ef.id, aid })))
    .sort((a, b) => b.q - a.q || a.n - b.n || cmp(b.eid, a.eid) || cmp(b.aid, a.aid));
  const usedE = new Set<string>();
  const usedA = new Set<string>();
  const match: Record<string, string> = {};
  for (const p of pairs) {
    if (usedE.has(p.eid) || usedA.has(p.aid)) continue;
    usedE.add(p.eid);
    usedA.add(p.aid);
    match[p.eid] = p.aid;
  }

  const rows: string[][] = [];
  let credit = 0;
  let blkN = 0;
  let blkHit = 0;
  let sevOk = 0;
  let routeOk = 0;
  for (const ef of exp) {
    const aid = match[ef.id];
    const af = aid ? actual.get(aid) : undefined;
    if (ef.min_severity === "blocker") blkN++;
    if (!af) {
      rows.push([ef.id, "—", "missed", "", "", ef.about]);
      continue;
    }
    const t = tOk(ef, af);
    const s = sOk(ef, af);
    const r = rOk(ef, af);
    credit += t ? 1 : 0.5;
    sevOk += Number(s);
    routeOk += Number(r);
    if (ef.min_severity === "blocker" && s) blkHit++;
    rows.push([
      ef.id,
      aid,
      t ? "matched" : "found-mislabelled (unconfirmed)",
      `${af.severity}${s ? "" : ` ✕ (≥ ${ef.min_severity})`}`,
      `${af.route}${r ? "" : ` ✕ ${ef.routes.join("/")}`}`,
      ef.about,
    ]);
  }
  const nM = Object.keys(match).length;
  const extras = [...actual.keys()].filter((a) => !usedA.has(a));
  const m: SeededScore["metrics"] = {
    "recall.blocker": blkN ? round3(blkHit / blkN) : 1,
    "recall.all": round3(credit / exp.length),
    severity_accuracy: nM ? round3(sevOk / nM) : 0,
    route_accuracy: nM ? round3(routeOk / nM) : 0,
    noise: actual.size ? round3(extras.length / actual.size) : 0,
  };
  const checks = seededChecks(m, thresholds.release[step].seeded);
  const passed = checks.every(ok);
  const out = [
    `# Seeded score · ${c.id} · run \`${runName}\``,
    "",
    "| Expected | Actual | Result | Severity | Route | About |",
    "|---|---|---|---|---|---|",
    ...rows.map((r) => `| ${r.join(" | ")} |`),
    "",
    `Unmatched actual findings (label each \`valid-extra\` or \`noise\`): ${extras.join(", ") || "none"}`,
    "",
    `| Metric | Value | Release (${step}) |`,
    "|---|---|---|",
    ...checkRows(checks),
    "",
    `**Case result: ${passed ? "PASS" : "FAIL"}** (single run; release uses the worst of k=${thresholds.release.k_runs} for hard rules)`,
  ];
  return { data: { case: c.id, run: runName, metrics: m, matches: match, extras, pass: passed }, report: out.join("\n") };
}

// ---------------------------------------------------------------- §B golden (proxy)

export interface GoldenScore {
  case: string;
  run: string;
  metrics: { ref_coverage: number; must_miss: number };
  alignment: Record<string, [string, string]>;
  must_missed: string[];
  extras: string[];
  pass: boolean;
  proxy: true;
}

export function goldenChecks(m: GoldenScore["metrics"], th: Yaml): Check[] {
  return [
    { metric: "ref_coverage", value: m.ref_coverage, limit: th.ref_coverage, kind: "min" },
    { metric: "must_miss", value: m.must_miss, limit: th.must_miss, kind: "max" },
  ];
}

export function scoreGolden(caseDir: string, specDir: string, runName = basename(specDir), thresholds = loadThresholds()): { data: GoldenScore; report: string } {
  const c = loadCaseYaml(caseDir);
  const refDir = resolve(caseDir, c.expected.reference);
  const reqs = (dir: string) => new Map([...parseArtifact(join(dir, "requirements.md")).blocks].filter(([id, b]) => id.startsWith("REQ") && live(b)));
  const R = reqs(refDir);
  const A = reqs(specDir);
  const jac = (a: string[], b: string[]) => {
    const sa = new Set(a);
    const sb = new Set(b);
    const union = new Set([...sa, ...sb]);
    return union.size ? [...sa].filter((x) => sb.has(x)).length / union.size : 0;
  };
  const cmp = (a: string, b: string) => (a < b ? -1 : a > b ? 1 : 0);
  const cand: Array<{ j: number; rid: string; aid: string }> = [];
  for (const [rid, r] of R)
    for (const [aid, a] of A) {
      const j = jac(fieldList(r, "terms"), fieldList(a, "terms"));
      if (fieldScalar(r, "context") === fieldScalar(a, "context") && j >= 0.4) cand.push({ j, rid, aid });
    }
  // Equivalents first, then by Jaccard, ties to the lower reference ID.
  cand.sort((x, y) => Number(y.j >= 0.6) - Number(x.j >= 0.6) || y.j - x.j || cmp(x.rid, y.rid) || cmp(x.aid, y.aid));
  const usedR = new Set<string>();
  const usedA = new Set<string>();
  const res = new Map<string, { aid: string; decision: string; credit: number; j: number }>();
  for (const { j, rid, aid } of cand) {
    if (usedR.has(rid) || usedA.has(aid)) continue;
    usedR.add(rid);
    usedA.add(aid);
    res.set(rid, { aid, decision: j >= 0.6 ? "equivalent" : "partial", credit: j >= 0.6 ? 1 : 0.5, j });
  }
  for (const { j, rid, aid } of cand) {
    // Merges: an actual REQ already matched also covers this reference REQ equivalently.
    if (usedR.has(rid) || j < 0.6) continue;
    usedR.add(rid);
    res.set(rid, { aid, decision: "merged", credit: 0.5, j });
  }
  const prio = (b: Block) => fieldScalar(b, "priority") ?? "";
  const w = (b: Block) => PRIORITY_W[prio(b)] ?? 0;
  let numer = 0;
  let denom = 0;
  for (const [rid, r] of R) {
    numer += w(r) * (res.get(rid)?.credit ?? 0);
    denom += w(r);
  }
  const mustMissed = [...R].filter(([rid, r]) => prio(r) === "must" && !res.has(rid)).map(([rid]) => rid);
  const extras = [...A.keys()].filter((a) => !usedA.has(a));
  const m = { ref_coverage: denom ? round3(numer / denom) : 0, must_miss: mustMissed.length };
  const checks = goldenChecks(m, thresholds.release["req-generation"].golden);
  const passed = checks.every(ok);
  const out = [
    `# Golden score (proxy) · ${c.id} · run \`${runName}\``,
    "",
    "| Reference | Priority | Run | Decision (unconfirmed) | Term Jaccard |",
    "|---|---|---|---|---|",
    ...[...R].map(([rid, r]) => {
      const x = res.get(rid);
      return `| ${rid} · ${r.title} | ${prio(r)} | ${x?.aid ?? "—"} | ${x?.decision ?? "missed"} | ${(x?.j ?? 0).toFixed(2)} |`;
    }),
    "",
    `Run REQs with no reference match (review: \`valid-extra\` / \`scope-creep\` / \`duplicate\`): ${extras.map((a) => `${a} · ${A.get(a)!.title}`).join(", ") || "none"}`,
    "",
    "| Metric | Value | Release (req-generation) |",
    "|---|---|---|",
    ...checkRows(checks),
    "",
    `**Case result (proxy): ${passed ? "PASS" : "FAIL"}**. The judge confirms or overturns each decision, then scores \`ac_coverage\`.`,
  ];
  const alignment = Object.fromEntries([...res].map(([rid, x]) => [rid, [x.aid, x.decision] as [string, string]]));
  return { data: { case: c.id, run: runName, metrics: m, alignment, must_missed: mustMissed, extras, pass: passed, proxy: true }, report: out.join("\n") };
}

// ---------------------------------------------------------------- judge fixture check

export interface JudgeScore {
  case: string;
  verdict_accuracy: number;
  verdict_ok: boolean;
}

export function scoreJudge(caseDir: string, judgeJson: string, thresholds = loadThresholds(), evals = EVALS): { data: JudgeScore; report: string } {
  const c = loadCaseYaml(caseDir);
  const out = JSON.parse(readFileSync(judgeJson, "utf8")) as Yaml;
  const validate = new Ajv2020({ allErrors: true, strict: false }).compile(JSON.parse(readFileSync(join(evals, "schemas", "judge-output.schema.json"), "utf8")));
  if (!validate(out)) throw new Error(`${judgeJson} does not match judge-output.schema.json: ${JSON.stringify(validate.errors)}`);
  const got = new Map<string, number>(out.criteria.map((x: Yaml) => [x.id, x.score]));
  let okN = 0;
  const rows: string[] = [];
  for (const ev of c.expected.judge_verdicts as Yaml[]) {
    const s = got.get(ev.criterion);
    const min = ev.min_score ?? 1;
    const hit = s !== undefined && min <= s && s <= ev.max_score;
    okN += Number(hit);
    rows.push(`| ${ev.criterion} | ${min}–${ev.max_score} | ${s ?? "missing"} | ${hit ? "✓" : "✕"} | ${ev.why ?? ""} |`);
  }
  const vOk = out.verdict === c.expected.verdict;
  const acc = round3((okN + Number(vOk)) / (rows.length + 1));
  const report = [
    `# Judge fixture · ${c.id}`,
    "",
    "| Criterion | Expected | Judge | OK | Why |",
    "|---|---|---|---|---|",
    ...rows,
    "",
    `Verdict: expected **${c.expected.verdict}**, judge said **${out.verdict}** ${vOk ? "✓" : "✕"}`,
    "",
    `\`verdict_accuracy\` = ${num(acc)} (release ≥ ${num(thresholds.release["eval-judge"].judge.verdict_accuracy)})`,
  ];
  return { data: { case: c.id, verdict_accuracy: acc, verdict_ok: vOk }, report: report.join("\n") };
}
